node/overseer/overseer-gen/proc-macro/src/impl_builder.rs

```diff
@@ -107,49 +81,256 @@ pub(crate) fn impl_builder(info: &OverseerInfo) -> proc_macro2::TokenStream {
+						F: 'static + FnOnce(#handle) ->
+							::std::result::Result<#field_type, #error_ty>,
+					{
+						let boxed_func = Init::<#field_type>::Fn(
+							Box::new(subsystem_init_fn) as SubsystemInitFn<#field_type>
+						);
+						#builder {
+							#field_name: boxed_func,
+							#(
+								#to_keep_subsystem_name: self. #to_keep_subsystem_name,
+							)*
+							#(
+								#baggage_name: self. #baggage_name,
+							)*
+							spawner: self.spawner,
+						}
+					}
+				}
 
-	let subsyste_ctx_name =
-		Ident::new(&(overseer_name.to_string() + "SubsystemContext"), overseer_name.span());
+				impl <InitStateSpawner, #field_type, #( #impl_subsystem_state_generics, )* #( #baggage_passthrough_state_generics, )*>
+				#builder <InitStateSpawner, #( #post_setter_state_generics, )* #( #baggage_passthrough_state_generics, )*>
+				where
+					#field_type : Subsystem<#subsystem_ctx_name<#subsystem_consumes>, #error_ty>,
+				{
+					/// Replace a subsystem by another implementation for the
+					/// consumable message type.
+					pub fn #field_name_replace<NEW, F>(self, gen_replacement_fn: F)
+						-> #builder <InitStateSpawner, #( #post_replace_state_generics, )* #( #baggage_passthrough_state_generics, )*>
+					where
+						#field_type: 'static,
+						F: 'static + FnOnce(#field_type) -> NEW,
+						NEW: #support_crate ::Subsystem<#subsystem_ctx_name< #subsystem_consumes >, #error_ty>,
+					{
+						let replacement: Init<NEW> = match self.#field_name {
+							Init::Fn(fx) =>
+								Init::<NEW>::Fn(Box::new(move |handle: #handle| {
+								let orig = fx(handle)?;
+								Ok(gen_replacement_fn(orig))
+							})),
+							Init::Value(val) =>
+								Init::Value(gen_replacement_fn(val)),
+						};
+						#builder {
+							#field_name: replacement,
+							#(
+								#to_keep_subsystem_name: self. #to_keep_subsystem_name,
+							)*
+							#(
+								#baggage_name: self. #baggage_name,
+							)*
+							spawner: self.spawner,
+						}
+					}
+				}
+			}
+		});
+
+	// Produce setters for all baggage fields as well
+	let baggage_specific_setters = info.baggage().iter().enumerate().map(|(idx, bag_field)| {
+		// Baggage fields follow subsystems
+		let fname = &bag_field.field_name;
+		let field_type = &bag_field.field_ty;
+		let impl_baggage_state_generics = recollect_without_idx(&baggage_passthrough_state_generics[..], idx);
+		let to_keep_baggage_name = recollect_without_idx(&baggage_name[..], idx);
+
+		let mut pre_setter_generics = baggage_passthrough_state_generics
+			.iter()
+			.map(|gen_ty| parse_quote!(#gen_ty))
+			.collect::<Vec<syn::GenericArgument>>();
+		pre_setter_generics[idx] = parse_quote! { Missing<#field_type> };
+
+		let mut post_setter_generics = pre_setter_generics.clone();
+		post_setter_generics[idx] = parse_quote! { Init<#field_type> };
+
+		// Baggage can also be generic, so we need to include that to a signature
+		let additional_baggage_generic = if bag_field.generic {
+			quote! {#field_type,}
```

Comment:
nit: remove the `,` here..

Reply:
nit: also a misnomer, it's not additional, it has been there from the get go :)

Reply:
Unfortunately, `,` can not be removed from here, as the `else` branch might have an empty token leading to double `,` syntax error. Renaming the variable is a good idea anyway.

Reply:
It could, but there is not much gain in this context, after all - discard my comment